### PR TITLE
Configure command parameters as nullable for optional dependents in table splitting

### LINKED ISSUE #21612: Table Splitting Discriminator property should be nullable by default
I ran into a problem while validating Can_manipulate_entities_sharing_row_independently in TableSplittingTestBase. The test attempts to remove one of the entities ("Engine") from a shared table ("Vehicles"), which really just sets two columns ("Description" and "Engine_Discriminator") to NULL. This fails with our ADO.NET provider, because the `DbParameter.Value` for the discriminator column is `DbNull.Value`, but `DbParameter.IsNullable` is `false`.

If I'm understanding the Table Splitting feature correctly, the discriminator properties should all be nullable by default. The TypeMappedRelationalParameter.AddParameter implementation creates a new instance of the relational parameter, which inherits IsNullable from property.IsNullable. Eventually this becomes a provider-specific implementation of DbParameter with IsNullable=false.

Please confirm if this is a bug in EF Core, or perhaps in the Functional Test (the Metadata.IsNullable could be set to true for the discriminator), or if I'm mistaken in my understanding of the feature.

### Further technical details

EF Core version: 3.1.5
Database provider: the one I'm writing

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #21612

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
@msftbot Require 1 sign off

File: src/EFCore.Relational/Storage/Internal/TypeMappedPropertyRelationalParameter.cs [hunk 1]
```diff
@@ -14,6 +15,7 @@ namespace Microsoft.EntityFrameworkCore.Storage.Internal
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
+    [Obsolete]
     public class TypeMappedPropertyRelationalParameter : TypeMappedRelationalParameter
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/Internal/TypeMappedPropertyRelationalParameter.cs:19]: .Internal. Can remove it.

File: src/EFCore.Relational/Storage/TypeMaterializationInfo.cs [hunk 2]
```diff
@@ -54,6 +56,30 @@ public class TypeMaterializationInfo
...
+        ///     Creates a new <see cref="TypeMaterializationInfo" /> instance.
+        /// </summary>
+        /// <param name="modelClrType"> The type that is needed in the model after conversion. </param>
+        /// <param name="property"> The property associated with the type, or <see langword="null" /> if none. </param>
+        /// <param name="mapping"> The type mapping to use or <see langword="null" /> to infer one. </param>
+        /// <param name="nullable"> A value indicating whether the value could be null. </param>
+        public TypeMaterializationInfo(
+            [NotNull] Type modelClrType,
+            [CanBeNull] IProperty property,
+            [NotNull] RelationalTypeMapping mapping = null,
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/TypeMaterializationInfo.cs:69]: NotNull attribute and default value null?

File: test/EFCore.SqlServer.FunctionalTests/DataAnnotationSqlServerTest.cs [hunk 3]
```diff
@@ -220,64 +220,6 @@ public override void MaxLengthAttribute_throws_while_inserting_value_longer_than
...
-INSERT INTO [Sample] ([MaxLengthProperty], [Name], [RowVersion], [AdditionalDetails_Name], [Details_Name])
-VALUES (@p0, @p1, @p2, @p3, @p4);
-SELECT [UniqueNo]
-FROM [Sample]
-WHERE @@ROWCOUNT = 1 AND [UniqueNo] = scope_identity();");
-        }
-
-        public override void RequiredAttribute_for_navigation_throws_while_inserting_null_value()
-        {
-            base.RequiredAttribute_for_navigation_throws_while_inserting_null_value();
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/DataAnnotationSqlServerTest.cs:233]: What happened to these assertions?

File: src/EFCore.Relational/Storage/TypeMaterializationInfo.cs [hunk 4]
```diff
@@ -24,6 +24,7 @@ public class TypeMaterializationInfo
         ///     The index of the underlying result set that should be used for this type,
         ///     or -1 if no index mapping is needed.
         /// </param>
+        [Obsolete]
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/TypeMaterializationInfo.cs:27]: Now query does not use this part, may be we should clean it up in some future release.

**Comment by smitpatel** [comment]
I was late! 😞

**Comment by ajcvickers** [comment]
@smitpatel Don't think you can compete against the robots and win! Bad mistake.

File: src/EFCore.Relational/Storage/Internal/TypeMappedPropertyRelationalParameter.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Storage/Internal/TypeMappedPropertyRelationalParameter.cs:19]: It's being used by `RelationalCommandBuilderExtensions.AddPropertyParameter`

File: src/EFCore.Relational/Storage/TypeMaterializationInfo.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Storage/TypeMaterializationInfo.cs:69]: If only there was some feature in C# that would show an error when a null is assigned to a non-nullable reference type...

File: test/EFCore.SqlServer.FunctionalTests/DataAnnotationSqlServerTest.cs (hunk 3 above)
Comment (AndriySvyryd) [reply test/EFCore.SqlServer.FunctionalTests/DataAnnotationSqlServerTest.cs:233]: The error is thrown before the query is executed, so we don't need to assert SQL

File: src/EFCore.Relational/Storage/TypeMaterializationInfo.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Storage/TypeMaterializationInfo.cs:27]: Yes, that's why I put Obsolete on all of these

**Pull request merged**